lib/orbit/rail_line.ex

```diff
@@ -0,0 +1,20 @@
+defmodule Orbit.RailLine do
+  use Ecto.Type
+
+  @type t :: :red | :orange | :blue
+
+  @impl true
+  def type, do: :string
+
+  @impl true
+  def cast(line) when line in [:red, :orange, :blue], do: {:ok, line}
```

Comment:
Is there no need for a cast from string?

Reply:
The [`cast/1` callback](https://hexdocs.pm/ecto/Ecto.Type.html#c:cast/1) here is used as part of the Ecto changeset "cast" operation, it basically just validates. Actually turning it into the value that gets sent to the DB is the job of the `dump/1` callback.

Reply:
Gotcha. Since it's effectively an identity function, could it be omitted?

Reply:
It's a required callback of the [`Ecto.Type`](https://hexdocs.pm/ecto/Ecto.Type.html#c:cast/1) behaviour, so if we don't define it, Elixir will complain. It's an identity when actually given one of the allowed values, but its real purpose is populating a useful error message if appropriate.